### PR TITLE
Set `-DHAVE_USLEEP` to solve poor concurrency on Unix

### LINKED ISSUE #280: Stabilize parallel writes to database tables
I would like to utilize sqlite3_busy_handler() and squlite3_busy_timeout() to check if a database is locked and to retry RSQLite::dbWriteTable once the database is no longer locked. Is there any implementation of this in RSQLite?

### PR DESCRIPTION
**Pull request by gaborcsardi**
Without this, sqlite3 can only wait whole seconds when waiting on a locked DB, which greatly increases the chance that processes time out. @Enchufa2 found a blog post that describes the issue in detail: https://beets.io/blog/sqlite-nightmare.html

`usleep()` is not used on Windows, and it is present on every Unix system. It is actually considered to be obsolete, in favor of `nanosleep()`, but systems still have it for compatibility.

Also added the `dbSetBusyHandler()` function to expose the `sqlite3_busy_timeout()` and `sqlite3_busy_handler()` APIs.

The new test case will fail for the first commit of the PR (on Unix), this will be fixed in subsequent commits.

The main purpose of adding `dbSetBusyHandler()` is to debug concurrency issues, and to show that adding the `-DHAVE_USLEEP` compile time flag (coming in the next commit) indeed improves the concurrency of RSQLite a lot.

See #280, #56, https://github.com/r-lib/liteq/issues/28, etc.

Closes #280.

### CODE REVIEW

**Comment by krlmlr** [comment]
Thanks for digging through this, looks great!

- Can we please make this a plain old function with `sqlite` prefix, e.g. `sqliteSetBusyHandler()`?
- The tests are still failing on the Mac.

**Comment by gaborcsardi** [comment]
> Can we please make this a plain old function with sqlite prefix, e.g. sqliteSetBusyHandler()?

Yep, that makes things a bit simpler.

> The tests are still failing on the Mac.

Which is pretty puzzling, because I wrote this mostly on macOS. :) I'll debug this.

**Comment by krlmlr** [comment]
Can you please rename, then I'll merge and cran.

**Comment by gaborcsardi** [comment]
> Can you please rename, then I'll merge and cran.

Thanks, let's make sure this is all correct. :) E.g. it would be great to hear from people that setting a high enough busy timeout solves their issues. 

I also added more docs, can you please check quickly if they make sense?

Btw. I tested this on Solaris as well, but will test that last version again...

**Comment by gaborcsardi** [comment]
@krlmlr Also, please let me know is this is what you meant by a "plain old function".

**Comment by gaborcsardi** [comment]
@krlmlr These timing based tests are always somewhat fragile. I don't think these are too bad, but in theory they might still fail if the system very-very busy. Should we skip them on CRAN?

**Comment by krlmlr** [comment]
Yeah, let's skip on CRAN.

**Review by krlmlr — COMMENTED** [review]
Thanks, the docs are great.

Should we allow the callback to throw errors (or at least catch them) so that the user can interrupt? How do we test?

File: src/DbConnection.cpp [hunk 1]
```diff
@@ -20,6 +21,8 @@ DbConnection::~DbConnection() {
   if (is_valid()) {
     disconnect();
   }
+  // in case this is still lingering for an invalid connection
+  release_callback_data();
```
[the remark this answers is not in the document]
Comment (gaborcsardi) [reply src/DbConnection.cpp:25]: This is the destructor.

[the remark this answers is not in the document]
Comment (krlmlr) [reply src/DbConnection.cpp:25]: Yeah, I was fooled by the collapsed lines.

**Comment by krlmlr** [comment]
Maybe catch errors from the callback with `try {}` and (later) propagate them to the original caller?

**Comment by gaborcsardi** [comment]
> Maybe catch errors from the callback with try {} and (later) propagate them to the original caller?

I tested it a bit, and apparently the connection is still fine after the interruption or error in the callback. I can add a test case for it as well. 

So maybe we just document that this is what we see and that a callback function is currently experimental. Most people are fine with the timeout, anyway.

**Comment by krlmlr** [comment]
Sounds great!

**Comment by gaborcsardi** [comment]
OK, added a note about errors/interruption.

**Comment by krlmlr** [comment]
Thanks!

**Comment by krlmlr** [comment]
Winbuilder appears stuck, might hit some time until I can actually submit to CRAN.

**Comment by gaborcsardi** [comment]
> Winbuilder appears stuck, might hit some time until I can actually submit to CRAN.

The CI is fine on Windows, so you could also just submit? The incoming checks will check it on Windows as well, anyway.

**Pull request merged**